### ISSUE TITLE
Possible to continue pretraining?

### ISSUE THREAD

**Issue by rohan-mehta** [facebookresearch/dinov2#204]
I know the general recommendation is to leave the backbone frozen and train task-specific heads. However I'm interested in continuing pre-training to better fit the backbone features to my dataset. Is this feasible?

I did try it, with the steps below, but found that KNN performance started to degrade very quickly - after 20-30K images, it dropped below 10%.
My approach:
1. Load the provided backbone weights into the `[SSLMetaArch](https://github.com/facebookresearch/dinov2/blob/main/dinov2/train/ssl_meta_arch.py#L31)` model. This provides weights for everything except the dino/ibot heads, so train just those heads on a few 100K samples (i.e. rest of the model is frozen). At the end of this, the model still has good KNN performance.
2. Unfreeze the entire model, and run `train.py` on ImageNet-1K (I first tried on my own dataset, but switched to ImageNet just to make sure its not a dataset issue). When I do this, the performance degrades quickly. After 10K images, KNN performance goes from 80%->25%. After another 10K, its ~10%. And it keeps dropping like that.

I'm using the standard VIT-L config provided in the repo.

So I'm looking for any advice or thoughts here - am I approaching this wrong? Or is this somehow not possible? Thanks!

**Comment by qasfb** [comment]
So this is not possible because we don't provide the training heads at the moment. We haven't explored this case in depth so I can't tell you what would be the right way, but we may plan to do it in the future.

**Comment by rohan-mehta** [comment]
@qasfb the codebase does have a “dino_head” and a “ibot_head”. Are those not the training heads? 

And as mentioned I first trained just those heads

**Comment by tZimmermann98** [comment]
@rohan-mehta did you manage to make some progress on this in between? Currently facing the same issue and would like to continue pre-training to adapt to the domain of my use case. How long did you kept the Backbone frozen in the first step?

**Comment by csaroff** [comment]
@qasfb fwiw, we'd be really interested in doing this as well!

**Comment by echochoc** [comment]
In my case, my training dataset will periodically grow over time, continuously incorporating images from new categories. I hope to train the model with long-life learning. Any ideas?

**Comment by Badar-e-Alam** [comment]
@csaroff @MarioAvolio @rohan-mehta , Here's what I did: Initially, I froze the feature extractor and solely focused on training the classifier head. This approach yielded approximately 85% accuracy. However, I then decided to fine-tune both the feature extractor and the classifier head, resulting in an 8% increase in accuracy. Just to give you some context, I'm working with a custom dataset comprising roughly 38,000 samples.

**Comment by vanpersie32** [comment]
@Badar-e-Alam Thank you Alam. I am doing continue pretraining on my custom dataset. May I have your Facebook Id, I want to chat with you.

**Comment by dimidagd** [comment]
> @Badar-e-Alam Thank you Alam. I am doing continue pretraining on my custom dataset. May I have your Facebook Id, I want to chat with you.

Hi @vanpersie32 can we get in contact somehow? :)

**Comment by vanpersie32** [comment]
> > @Badar-e-Alam Thank you Alam. I am doing continue pretraining on my custom dataset. May I have your Facebook Id, I want to chat with you.
> 
> Hi @vanpersie32 can we get in contact somehow? :)

sorry I didn't notice the message, my email is 18810388176@163.com.Hope to contact with you.

**Comment by Badar-e-Alam** [comment]
@vanpersie32  Apologies for the late response. Please check your email.

**Comment by wheel-is** [comment]
https://github.com/csaroff/dinov2/tree/main/sky

**Comment by Badar-e-Alam** [comment]
https://github.com/Badar-e-Alam/DINOv2_Downstream

**Comment by divyanshj16** [comment]
@rohan-mehta did you use the default model logging to track your training?

**Comment by divyanshj16** [comment]
@rohan-mehta did you had to manually evaluate the model for KNN, I do not see any automatic KNN evaluation in the training script.

**Comment by dimidagd** [comment]
Yes, you can evaluate manually

**Comment by divyanshj16** [comment]
@dimidagd 
One thing I am unable to easily do is load pretrained checkpoint weights in the SSL_meta_arch student backbone.
I got the weights from torch hub, but the keys don't match. I expected them to match.
